agent/reporter.go

```diff
@@ -102,27 +144,25 @@ func (sr *statusReporter) run(ctx context.Context) {
 			return
 		}
 
-		for taskID, status := range sr.statuses {
-			delete(sr.statuses, taskID) // delete the entry, while trying to send.
-
-			sr.mu.Unlock()
-			err := sr.reporter.UpdateTaskStatus(ctx, taskID, status)
-			sr.mu.Lock()
-
-			// reporter might be closed during UpdateTaskStatus call
-			if sr.closed {
-				return
-			}
-
-			if err != nil {
-				log.G(ctx).WithError(err).Error("status reporter failed to report status to agent")
-
-				// place it back in the map, if not there, allowing us to pick
-				// the value if a new one came in when we were sending the last
-				// update.
-				if _, ok := sr.statuses[taskID]; !ok {
-					sr.statuses[taskID] = status
-				}
+		// swap out the statuses map, so we can release the lock while we
+		// do the batch
+		statuses := sr.statuses
+		sr.statuses = map[string]*api.TaskStatus{}
+		// unlock the map so new statuses can be added while we process the
+		// current ones
+		sr.mu.Unlock()
+		err := sr.reporter.UpdateTaskStatuses(ctx, statuses)
+		// re-lock the map so we can add statuses back if we need to.
+		sr.mu.Lock()
```

Comment:
Non-blocking, just confirming my understanding: although it might be possible for the status reporter to be closed while `UpdateTaskStatuses` is executing, we don't care because adding statuses to a closed reporter won't cause any issues?